cow_instrument/DetectorInfo.h

```diff
@@ -0,0 +1,190 @@
+#ifndef DETECTOR_INFO_H
+#define DETECTOR_INFO_H
+
+#include <memory>
+#include <vector>
+#include <stdexcept>
+#include <string>
+#include <sstream>
+#include <cmath>
+#include <boost/optional.hpp>
+#include "Detector.h"
+#include "IdType.h"
+#include "MaskFlags.h"
+#include "MonitorFlags.h"
+#include "L2s.h"
+#include "Command.h"
+#include "Spectrum.h"
+#include "cow_ptr.h"
+#include <mutex>
+#include <thread>
+
+namespace {
+
+template <typename U>
+void detectorRangeCheck(size_t detectorIndex, const U &container) {
+  if (detectorIndex >= container.size()) {
+    std::stringstream buffer;
+    buffer << "Detector at index " << detectorIndex << " is out of range";
+    throw std::out_of_range(buffer.str());
+  }
+}
+
+double distance(const V3D &a, const V3D &b) {
+  return std::sqrt(((a[0] - b[0]) * (a[0] - b[0])) +
+                   ((a[1] - b[1]) * (a[1] - b[1])) +
+                   ((a[2] - b[2]) * (a[2] - b[2])));
+}
+}
+
+template <typename InstTree> class DetectorInfo {
+public:
+  template <typename V>
+  explicit DetectorInfo(V &&instrumentTree)
+      : m_nDetectors(instrumentTree->nDetectors()),
+        m_isMasked(std::make_shared<MaskFlags>(m_nDetectors, Bool(false))),
+        m_isMonitor(std::make_shared<MonitorFlags>(m_nDetectors, Bool(false))),
+        m_l2(std::make_shared<L2s>(m_nDetectors)),
+        m_sourcePos(instrumentTree->sourcePos()),
+        m_samplePos(instrumentTree->samplePos()),
+        m_instrumentTree(instrumentTree) {
+
+    /* The lifetime of the DetectorInfo is tied to the lifetime of the
+       InstrumentTree (see constructor). So no danger that instrument movement
+       gets
+       out of step with this. InstrumentTrees are immutable.
+     */
+
+    // Calculate this once.
+    m_l1 = distance(m_sourcePos, m_samplePos);
+    initL2();
+  }
+
+  // template <typename V>
+  // DetectorInfo(const DetectorInfo<V> &other) = default;
+
+  template <typename V>
+  std::unique_ptr<DetectorInfo> cloneWithInstrumentTree(V &&instrumentTree) {
```

Comment:
I don't quite like this method, the mechanism of replacing the tree feels a bit weird.